### PR TITLE
Breaking: Error thrown/printed if no config found (fixes #5987)

### LINKED ISSUE #5987: Proposal: throw error if no config file is found
From my travels around the internet, it seems that first-time ESLint users are very confused about the "nothing by default" mode. It seems many do not read the getting started docs, and so don't know to use `--init` to get started. That leaves them thinking that ESLint is broken because JSHint shows output out of the box. 

~~To remedy this, I'd like to propose that in 3.0, we change the default behavior to use `eslint:recommended ` when we don't find a config file. This should only affect CLI users and not affect `CLIEngine` by default (so as not to affect API consumers).~~

~~Another option is to automatically run `--init` when a config file isn't found, but that won't work well in editors, where an interactive prompt won't be available.~~

**Edit:** new proposal is throwing an error asking the user to run --init (we can use a pretty error message via messagsTemplate rather than showing the stack trace).

### PR DESCRIPTION
**Pull request by platinumazure**
Pulled in @IanVS' commit and added a few of my own to reorganize things a bit. I'm *fairly* sure this is ready, but let me know if I can organize things better (or if I should revert some of my last commits).

Organization stuff:
* Created new sub-describe in tests for cases where local and personal config does not exist (throws unless no-eslintrc or unless there is at least one rule key/value pair)
* Tried to make lib/config.js getLocalConfig() have fewer exit points. At the end is an if statement setting up the throw condition, and immediately after that is an unconditional return.

Thanks in advance for any feedback.

Ping @IanVS, @nzakas

### CODE REVIEW

File: lib/config.js [hunk 1]
```diff
@@ -140,8 +150,35 @@ function getLocalConfig(thisConfig, directory) {
...
+         * - Otherwise, throw an exception.
+         */
+        personalConfig = getPersonalConfig();
+
+        if (personalConfig) {
+            config = ConfigOps.merge(config, personalConfig);
+        }
+    }
+
+    if (!found && !thisConfig.useSpecificConfig && !personalConfig && !optionsHaveRules(thisConfig.options)) {
```
Comment (IanVS) [inline lib/config.js]: I wonder if this wouldn't be more clear by moving it to `else if (!optionsHaveRules(thisConfig.options))` after the `if (personalConfig)` above.

**Comment by IanVS** [comment]
One super-minor comment, but LGTM otherwise.

**Comment by platinumazure** [comment]
I'm willing to try reconfiguring it, but I intentionally chose this way
(knowing it is redundant) for readability. With the current layout, you
know exactly why we're throwing the exception and what we've tried.

Still want me to change it? I'm open to it.

File: lib/config.js [hunk 2]
```diff
@@ -87,7 +87,16 @@ function getPersonalConfig() {
...
     }
 
-    return config || {};
+    return config || null;
+}
+
+/**
+ * Determine if rules were explicitly passed in as options.
+ * @param {Object} options The options used to create our configuration.
+ * @returns {bool} True if rules were passed in as options, false otherwise.
```
Comment (nzakas) [inline lib/config.js]: {boolean}

File: lib/config.js [hunk 3]
```diff
@@ -87,7 +87,16 @@ function getPersonalConfig() {
...
+ */
+function optionsHaveRules(options) {
```
Comment (nzakas) [inline lib/config.js]: Functions should begin with verbs, so maybe `hasRules` instead?

File: lib/config.js (hunk 1 above)
Comment (nzakas) [inline lib/config.js]: Agree with @ianvs. Checking the same conditions twice is confusing.

File: messages/no-config-found.txt [hunk 4]
```diff
@@ -0,0 +1,7 @@
+ESLint couldn't find a configuration file. To set up a configuration file for this project, please run:
+
+    eslint --init
+
+We looked for configuration files in <%= directory =%> and its ancestors.
```
Comment (nzakas) [inline messages/no-config-found.txt]: We don't use "we" in other messages, use "ESLint" instead for consistency.

File: tests/lib/config.js [hunk 5]
```diff
@@ -103,6 +103,17 @@ describe("Config", function() {
         return path.join.apply(path, args);
     }
 
+    /**
+     * Mocks the current CWD path
+     * @param {string} fakeCWDPath - fake CWD path
+     * @returns {undefined}
```
Comment (nzakas) [inline tests/lib/config.js]: returns {void}

File: messages/no-config-found.txt (hunk 4 above)
Comment (platinumazure) [inline messages/no-config-found.txt]: Acknowledged. Note to myself on the same line, should be `<%= directory %>`.

File: lib/config.js (hunk 3 above)
Comment (platinumazure) [inline lib/config.js]: :+1:

**Comment by nzakas** [comment]
Just to double-check, this throws only when all of these are true?

1. No project config file
1. No personal config file
1. No `--rule`
1. `useEslintrc` is true

File: lib/config.js (hunk 1 above)
Comment (platinumazure) [inline lib/config.js]: Hmm. I can see that. I'm leaning towards extracting to a few more functions. See my latest comment on the issue itself for how I plan to handle this.

**Comment by IanVS** [comment]
> Still want me to change it? I'm open to it.

The way you have it is more confusing to me because the logic is being duplicated and I don't expect that.  And also your comment `Otherwise, throw an exception.` is inside the first if block.  But, I don't feel super-strongly about it.

**Comment by platinumazure** [comment]
My thoughts/planned next steps:

Right now `getLocalConfig()` is violating single responsibility principle (assuming "local config" refers specifically to a config in the same or ancestor directory, as contrasted with "personal config" for the home directory case). It shouldn't be responsible for finding the personal config as well.

I'm going to create an orchestration function that will hopefully make things easier to read overall. `getLocalConfig()` will focus on local/ancestor configs, `getPersonalConfig()` will probably be untouched, and the new orchestration function will basically be a big if/else to make things easier to read and less redundant.

**Comment by platinumazure** [comment]
@IanVS I can definitely improve the comments. But please do let me know your thoughts on [my previous comment](https://github.com/eslint/eslint/pull/6538#issuecomment-228795826)?

**Comment by nzakas** [comment]
@platinumazure I don't think you need to split everything into functions - just don't duplicate the `if` condition. I'd really like to get this merged ASAP so we have a couple of days before the release to test on master, so let's take the path of least resistance. You can always refactor later.

**Comment by platinumazure** [comment]
@nzakas @IanVS Updated with all feedback. I think this should be good to merge now.

@nzakas:

> Just to double-check, this throws only when all of these are true?
> 1. No project config file
> 1. No personal config file
> 1. No --rule
> 1. useEslintrc is true

Yes. In `getLocalConfig()`, if we don't find project config, we fall back to personal config file; and if both of those are missing, we check for rules having been passed in via `--rule`; and after all that, we throw. In the current implementation, `getLocalConfig()` isn't called if `useEslintrc` is false, so no throwing in that case either.

I'd still like to take a pass through the file to refactor for clarity, after 3.0.0 is out.

**Comment by nzakas** [comment]
LGTM. Nice work!

**Comment by platinumazure** [comment]
Thanks for your patience.

**Pull request merged**